Implement the lazy_load_members room state filter parameter

ara4n opened the pull request:
This is a first cut at filtering out room_members from sync responses unless they're actually needed to render the timeline (as proposed at https://docs.google.com/document/d/11yn-mAkYll10RJpN0mkYEVqraTbU3U4eQx9MNrzqX1U/edit#)

My hope is to get this merged so that client developers can experiment with lazy_loading and see how much it speeds up their clients, and to check how badly clients handle members trickling in on demand.

Must-have todo:
 * [x] Proactively add in new state events to responses as they're required.
 * [x] pep8
 * [x] pydoc
 * [x] write sytest
 * [x] Actually add API to turn the filtering on & off; atm it's hardcoded on
 * [x] opportunistically deduplicate members sent to clients within the same series of sync tokens (if the client wants it) - now done in #3331
 * [x] calculate room names serverside, given the client no longer has the full list of members in order to do it itself. (We might also want to revisit what these names look like, as "Matthew and 2 others" is a lot less useful than "Matthew, Tom, Amand..." or similar
 * [x] Handle disambiguating display names, given the client can't do it any more.
  * ...actually, the client /can/ disambiguate, as it should only ever need to disambiguate if it has actually seen some ambiguous members.  So even if in total there are ambiguous users in a room, if the client isn't aware of this, it shouldn't need to disambiguate, i think?
 * [x] supporting /messages -- we need this in order for historical displaynames & user_names to be correct at all.
 * [x] Some API needed to provide historical profile details when doing /context etc?
 * [x] Advise how to search the membership list. (do we just pull in /state without a filter for it?)
 * [x] test under postgres

Lower priority:
 * [ ] being smarter about which members are needed for a given block of timeline (atm we just pull in those for the senders of the events in the timeline, but need to consider invite targets etc too)
 * [x] supporting /context

turt2live commented:
Another thing that may need supporting is the context API if I'm not mistaken. Does this work as-is with Riot or should I hold off on deploying this to my test environment?

ara4n commented:
i haven’t tested it against riot yet, but in theory it should work. good call on /context. i would be *very* interested in a “this sped up initial sync from X to Y, and reduced the json from P to Q” stat.

ara4n commented:
now updated to try to always add in the necessary member events for a chunk of timeline (although sytest is reporting 500s so clearly needs more work)

ara4n commented:
So I just put this live with a custom sync worker against my `@matthew2:matrix.org` account.  matthew2's vital stats are:

```
total state events: 57227
total member events: 56085
total timeline events: 2479
total rooms: 135
possible speedup factor: 32.859658778205834x
```

Before (matrix-org-hotfixes branch)
 * cold initial sync: 36.7s
 * warm initial sync: 8.5s
 * sync size: 7197kB
 * v8 heap in riot/web: 252MB

After (matthew/filter_members branch)
 * cold initial sync: 12.0s
 * warm initial sync: 5.2s
 * sync size: 1746k
 * v8 heap after: 110MB

So, it looks like (with this impl at least) we're seeing a roughly 2-3x improvement on various metrics.

turt2live commented:
It looks like this isn't sending the senders of events on incremental syncs. It's not the end of the world yet (as it doesn't actually break anything as far as I can tell), it just looks bad in Riot.

(I also realize this isn't near complete yet - just lodging the bug now for consideration)

ara4n commented:
yeah, incremental syncs are borked. my next step here is to write some sytests to try to get some level of confidence it actually works properly.

richvdh commented:
closing this for now. I don't think that PRs are a great place to track long-term proofs of concept.

turt2live commented:
Fixes https://github.com/matrix-org/synapse/issues/1203

turt2live commented:
Fixes https://github.com/matrix-org/synapse/issues/1379

ara4n commented:
I've added an initial sytest for this over at https://github.com/matrix-org/sytest/pull/445 which exercises initial sync.  It still needs one for incremental sync, gapped incremental sync & deduplicating redundant membership (when that gets there).

This should be good enough for a first cut, at least for clientside devs to develop against. The main missing feature is deduplicating redundant membership data - i.e. every /sync currently will return all the members required for its response, even if it just sent the same event IDs to the client in the previous /sync.  However, this is an (easy) optimisation, and doesn't impact client devs.

@richvdh PTAL, in the hope that this can be unblocked and free up the client devs to experiment with the lazy_load_members filter flag and ensure their clients cope okay with memberships being lazy.

ara4n commented:
@matrixbot retest this please

ara4n commented:
(btw as far as i can see the non-merged PR test fails are bogus)

richvdh commented on synapse/storage/state.py:
```diff
@@ -185,7 +185,19 @@ def get_state_groups(self, room_id, event_ids):
+            groups(list[int]): list of state group IDs to query
+            types(list[str|None, str|None])|None: List of 2-tuples of the form
+                (`type`, `state_key`), where a `state_key` of `None` matches all
+                state_keys for the `type`. Presence of type of `None` indicates
```
> Presence of type of `None` indicates that types not in the list should not be filtered out

I'm afraid I totally failed to grok this (which you have dutifully copied throughout the PR, thank you) without reading the code. I think it needs more explanation: "An entry of `(None, None)` is treated as a special case which means that any state events whose `type` does not appear anywhere else in the list will be included in the returned value", or words to that effect.

I'm also led to wonder if we shouldn't just make `include_other_types` another separate argument rather than this rather obscure special case.

Finally, I'm concerned that this API has rather surprising behaviour. For instance, suppose I wanted to get the list of all state events, except `m.room.member` which I want to filter by a list of users:

```
def get_limited_state(groups, users):
    _get_state_groups_from_groups(
        groups,
        [('m.room.member', u) for u in users],
        include_other_types=True,
   )
```

Now suppose I call that with an empty set of `users`. The expected behaviour is to get *no* `m.room.member` events; the *actual* behaviour is that I get *all* of them. Synapse has enough footguns without adding more.

An ideal solution might be for `types` to instead be a `dict[str, iterable[str]|None]` mapping from `type` to a set of state_keys, but that might require quite a lot of refactoring throughout the codebase. As a compromise: instead of the `(None, None)` special case (or a separate `include_other_types` param) add a `filtered_types` param (which would be set to `['m.room.member']` in the above usecase)?

ara4n replied:
have adopted `filtered_types` as suggested; it definitely feels better - thanks.

richvdh commented on synapse/storage/state.py:
```diff
@@ -469,20 +518,30 @@ def _get_some_state_from_cache(self, group, types):
         for typ, state_key in types:
             key = (typ, state_key)
+
+            if typ is None:
```
see the above comments about obscure special-cases, but if we're going to stick with this, could it be made to check for `(None, None)` rather than `(None, *)` for consistency with `_get_state_groups_from_groups`

ara4n replied:
fixed by `filtered_types`

richvdh commented on synapse/storage/state.py:
```diff
@@ -469,20 +518,30 @@ def _get_some_state_from_cache(self, group, types):
             if state_key is None:
                 type_to_key[typ] = None
+                # XXX: why do we mark the type as missing from our cache just
+                # because we weren't filtering on a specific value of state_key?
```
because when the cache was populated, it might have been done with a restricted set of `state_keys`, so the cache might be incomplete, I guess?

ara4n replied:
yup, i came to the same conclusion in the end. have updated the comment.

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -485,22 +495,60 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
+                    )
+                ]
+
+                # We can't remove redundant member types at this stage as it has
```
what are member types, and why would they be redundant?

ara4n replied:
this was a note to self which snuck in.

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -485,22 +495,60 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
                 if batch:
                     current_state_ids = yield self.store.get_state_ids_for_event(
-                        batch.events[-1].event_id
+                        batch.events[-1].event_id, types=types
```
trailing commas ftw in future, please

ara4n replied:
done

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -485,22 +495,60 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
         with Measure(self.clock, "compute_state_delta"):
+
+            types = None
+            member_state_ids = {}
```
could you add a comment explaining what this is? what does it map from and to?

richvdh replied:
given it's only used in the `full_state` and `batch.limited` branches below, it might make more sense to declare it in those branches.

ara4n replied:
done

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -485,22 +495,60 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
                     state_ids = current_state_ids
 
+                if lazy_load_members:
+                    member_state_ids = {
```
why are we doing this here and at line 582, rather than just doing it in `_calculate_state` based on a `lazy_load_members` flag?

richvdh replied:
oh, is it because of the TODO at line 578 - this is supposed to do something more intelligent?

ara4n replied:
yes. on the https://github.com/matrix-org/synapse/pull/3331/files branch the second branch evolves to filter out the redundant members to stop repeatedly spamming lazy-loaded members to the client in each and every /sync.

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -533,11 +592,23 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
             else:
                 state_ids = {}
+                if lazy_load_members:
+                    # TODO: filter out redundant members based on their mxids (not their
```
confused. What is a redundant member? Why do they have event_ids? are we currently filtering them by event_ids, or is this a note to remind us not to mistakenly filter by event_ids in the future?

ara4n replied:
have added a comment to answer the above.

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -533,11 +592,23 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
+                    # is an non-gappy incremental sync.
+
+                    # strip off the (None, None) and filter to just room members
+                    types = types[:-1]
```
rly? we've only just added the damn thing.

ara4n replied:
fixed by `filtered_types`

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -533,11 +592,23 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
+                    # event_ids) at this point. We know we can do it based on mxid as this
+                    # is an non-gappy incremental sync.
+
+                    # strip off the (None, None) and filter to just room members
```
whhhhyyyyy

ara4n replied:
fixed by `filtered_types`

ara4n commented on synapse/storage/state.py:
```diff
@@ -201,8 +213,18 @@ def _get_state_groups_from_groups(self, groups, types):
+                # for particular types.
+                include_other_types = True
+                type_set.remove((None, None))
+            types = list(type_set)  # deduplicate types list
```
cargo-culted from what was there before.

ara4n commented on synapse/storage/state.py:
```diff
@@ -249,6 +271,17 @@ def _get_state_groups_from_groups_txn(self, txn, groups, types=None):
+                    # XXX: check whether this slows postgres down like a list of
+                    # ORs does too?
+                    unique_types = set([t for (t, _) in types])
+                    clause_to_args.append(
```
done.

ara4n commented:
@richvdh thanks for the excellent review, and sorry it took so long to act on. I think i've incorporated all of your feedback, and added a UT for the state storage layer as requested.

ara4n commented:
(tests are failing because the sytest test is testing the next PR in this sequence: https://github.com/matrix-org/synapse/pull/3331)

ara4n commented:
@matrixbot retest this please

richvdh commented on synapse/storage/state.py:
```diff
@@ -185,23 +185,41 @@ def get_state_groups(self, room_id, event_ids):
+    ):
         results = {group: {} for group in groups}
+
+        include_other_types = False if filtered_types is None else True
```
I think it would be clearer to inline this (ie, `s/include_other_types/filtered_types is not None/` below)

ara4n replied:
done

richvdh commented on synapse/storage/state.py:
```diff
@@ -185,23 +185,41 @@ def get_state_groups(self, room_id, event_ids):
+            types(list[str|None, str|None])|None: List of 2-tuples of the form
+                (`type`, `state_key`), where a `state_key` of `None` matches all
+                state_keys for the `type`. If None, all types are returned.
+            filtered_types(list[str]|None): Only apply filtering via `types` to this
```
> Only apply filtering via `types`

no, this (and the subsequent implementation) misses the whole point of my suggestion here, which was based around the case of wanting to behave sensibly when `types` ended up empty.

The usecase I was worrying about was:
*  I want to get the list of *all* state events, except for `m.room.member`, which I want to filter by a list of users.
* Due to reasons, that list of users is empty.

So, we have:

```
store._get_state_groups_from_groups(groups, [], filtered_types=['m.room.member'])
```

Your code will return everything here (afaict). 

A correct description of how I envisaged this parameter would be:

> If not None, gives a list of state types which must match entries in `types` to be returned; other types of state are returned unfiltered. If None, *all* state types must match entries in `types`.

(which is still pretty confusing, and I welcome better suggestions. I do think that the current impl is a foot-gun, though.)

ara4n replied:
I seem to have a mental block on this; sorry.  The fact I can't understand what you're proposing makes me wonder how anyone else is going to understand it in future, unless this is just me being particularly thick.

As far as I can tell, my implementation and comment for `filtered_types` matches the behaviour that you're wanting.

```python
store._get_state_groups_from_groups(groups, [], filtered_types=['m.room.member'])
```

gets turned into:

```sql
select stuff where type <> 'm.room.member'
```

...which is what you were asking for?  (types is [], so you don't get any `and type=? and state_key=?` added to the where clause, leaving filtered_types, which is applied as a series of `and type <> ?` clauses).

Is the problem here that the python doc isn't clear enough?  Your suggestion of:

> If not None, gives a list of state types which must match entries in types to be returned; other types of state are returned unfiltered. If None, all state types must match entries in types.

seems to match the behaviour of my suggestion of:

> Only apply filtering via `types` to this list of event types.  Other types of events are returned unfiltered. If None, `types` filtering is applied to all events.

In other words: if `filtered_types` is specified, you should only apply filtering using the `types` field to the types given in the `filtered_types` list.  Others get returned unfiltered.  Is that clearer?

Or have I just got an enormous blindspot on this?

richvdh replied:
> ```python
> store._get_state_groups_from_groups(groups, [], filtered_types=['m.room.member'])
> ```
>
> gets turned into:
>
> ```sql
> select stuff where type <> 'm.room.member'
> ```

Does it? Are you sure? `[]` is falsy, so this block will get skipped: https://github.com/matrix-org/synapse/pull/2970/files#diff-6ee0c965f89a99e084e8500580e3ec17R258. Perhaps you could test this case in the UT.

>  The fact I can't understand what you're proposing makes me wonder how anyone else is going to understand it in future, unless this is just me being particularly thick.

I don't claim that what I'm suggesting is particularly elegant - as I say, I welcome better suggestions. I'm just trying to avoid this landing with the current massive footgun.

I think you just need to change the line I linked above (and its sqlite equivalent) to `if types is not None`.

ara4n replied:
oh, right - i was looking at the sqlite clause which gets it right already with `if types is not None:`; looks like i thinkoed with the `if types:` on the postgres clause. fixed.

ara4n commented on synapse/handlers/sync.py:
```diff
@@ -1472,11 +1564,12 @@ def _calculate_state(timeline_contains, timeline_start, previous, current):
+    tc_ids = set(e for e in timeline_contains.values())
 
-    state_ids = ((c_ids | ts_ids) - p_ids) - tc_ids
+    state_ids = (((c_ids | ts_ids) - p_ids) - tc_ids) | tsm_ids
```
hm, possibly not. i guess that if they are already in the timeline then it's redundant to also include them in state.  (although I'm surprised this didn't show up in sytests)

ara4n commented on synapse/storage/state.py:
```diff
@@ -185,23 +185,41 @@ def get_state_groups(self, room_id, event_ids):
 
-    def _get_state_groups_from_groups_txn(self, txn, groups, types=None):
+    def _get_state_groups_from_groups_txn(
+        self, txn, groups, types=None, filtered_types=None
```
done

ara4n commented on synapse/storage/state.py:
```diff
@@ -185,23 +185,41 @@ def get_state_groups(self, room_id, event_ids):
             res = yield self.runInteraction(
                 "_get_state_groups_from_groups",
-                self._get_state_groups_from_groups_txn, chunk, types,
+                self._get_state_groups_from_groups_txn, chunk, types, filtered_types
```
done

ara4n commented on synapse/storage/state.py:
```diff
@@ -185,23 +185,41 @@ def get_state_groups(self, room_id, event_ids):
+            types(list[str|None, str|None])|None: List of 2-tuples of the form
+                (`type`, `state_key`), where a `state_key` of `None` matches all
+                state_keys for the `type`. If None, all types are returned.
+            filtered_types(list[str]|None): Only apply filtering via `types` to this
```
done

ara4n commented on synapse/storage/state.py:
```diff
@@ -185,23 +185,41 @@ def get_state_groups(self, room_id, event_ids):
+
+        Args:
+            groups(list[int]): list of state group IDs to query
+            types(list[str|None, str|None])|None: List of 2-tuples of the form
```
done

ara4n commented:
@richvdh back with you.

ara4n commented on synapse/handlers/sync.py:
```diff
@@ -509,22 +562,46 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
+                    # members down at this point. Redundant members are ones which
+                    # repeatedly get sent down /sync because we don't know if the client
+                    # is caching them or not.
+                    member_state_ids = {
```
done

ara4n commented on synapse/handlers/sync.py:
```diff
@@ -485,22 +503,57 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
 
                     state_ids = current_state_ids
 
+                # track the membership state events as of the beginning of this
```
hopefully fixed whilst refactoring into _calculate_state.

ara4n commented on synapse/handlers/sync.py:
```diff
@@ -1472,11 +1564,12 @@ def _calculate_state(timeline_contains, timeline_start, previous, current):
+    tc_ids = set(e for e in timeline_contains.values())
 
-    state_ids = ((c_ids | ts_ids) - p_ids) - tc_ids
+    state_ids = (((c_ids | ts_ids) - p_ids) - tc_ids) | tsm_ids
```
i'm hoping the refactored solution is good enough, unless you're suggesting removing the intermediary sets on all of them?

ara4n commented on synapse/handlers/sync.py:
```diff
@@ -509,22 +562,46 @@ def compute_state_delta(self, room_id, batch, sync_config, since_token, now_toke
+                    filtered_types=filtered_types,
                 )
 
+                # track the membership state events as of the beginning of this
```
no, the idea is to just ensure that any members in the timeline get lazyloaded. we don't care what happened during the gap. hopefully moving it to _calculate_state has clarified things.

ara4n commented:
@richvdh ptal again.

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -1472,11 +1541,22 @@ def _calculate_state(timeline_contains, timeline_start, previous, current):
+    # so we can add them back in to the state if we're lazyloading.  We don't
+    # add them into state if they're already contained in the timeline.
+    if lazy_load_members:
+        ll_ids = set(
```
I wonder if this would be clearer rewritten as:

```
p.ids -= set (e for t, e in timeline_start.iteritems() if t[0] == EventTypes.Member)
```

or, more efficiently, especially when p_ids is empty, as it is for the initialsync case:

```
p_ids.difference_update(e for t, e in timeline_start.iteritems() if t[0] == EventTypes.Member)
```

ara4n replied:
i don't understand why you are trying to remove stuff from `p.ids`, which are the IDs of the state at the end of the previous sync (if any).  the idea here is to add state for the lazy-loaded members into the result, not remove stuff from the previous sync result, where it probably doesn't exist in the first place?

richvdh replied:
aren't the two equivalent?

ara4n replied:
ok.

richvdh commented on synapse/storage/state.py:
```diff
@@ -585,7 +649,7 @@ def _get_state_for_groups(self, groups, types=None):
                 types_to_fetch = types
```
it looks like this is going to do the wrong thing in the presence of an unwildcarded `types` but a non-empty `filtered_types`: in particular it isn't going to fetch the event types that aren't listed in `filtered_types`.

I'd say that, if filtered_types is not None, `types_to_fetch` should be set to None.

ara4n replied:
hopefully covered by my gist in the comment below, but if you can take a look at the gist patch before I put yet more dead ends into this PR it'd be appreciated.

richvdh commented on synapse/storage/state.py:
```diff
@@ -460,7 +511,7 @@ def _get_state_group_for_events(self, event_ids):
         defer.returnValue({row["event_id"]: row["state_group"] for row in rows})
 
-    def _get_some_state_from_cache(self, group, types):
+    def _get_some_state_from_cache(self, group, types, filtered_types=None):
```
I'm not sure this method does the right thing with filtered_types.

ara4n replied:
i've stared at this some more and am failing to guess or see why you think it's doing the wrong thing.

richvdh replied:
ok, suppose we have a partial cache of the state, including the `m.room.member` events, but not the `m.room.history_visibility` events. Suppose this is called with:

```
store._get_some_state_from_cache(group, [('m.room.member', 'user1')], filtered_types= ['m.room.member'])
```

We have all the events we need to satisfy the `types` list, so `missing_types` will be empty (and hence `got_all` will be True), even though we are missing important state.

ara4n replied:
right, missed that - thanks. hopefully e22700c3d resolves this.

richvdh replied:
ok, so now what happens when we have

```
store._get_some_state_from_cache(group, [], filtered_types= ['m.room.member'])
```

?

At this point I feel like I'm being a human unit-test rig, and it makes me sad.

ara4n replied:
sorry, i totally missed the messy [] special case. if it's any consolation, i strongly suspect i'd have missed it in a UT too.

richvdh replied:
I still don't think this is right; I don't think e22700c actually fixed the case I mentioned.

Prove me wrong with a test!

> i strongly suspect i'd have missed it in a UT too.

Well, can you add one now, please.

richvdh commented on synapse/storage/state.py:
```diff
@@ -585,7 +649,7 @@ def _get_state_for_groups(self, groups, types=None):
 
             group_to_state_dict = yield self._get_state_groups_from_groups(
-                missing_groups, types_to_fetch,
+                missing_groups, types_to_fetch, filtered_types
```
At the risk of being stabbed, now that you've done all the work to update `_get_state_groups_from_groups`...

I don't think we're ever going to get here with `filtered_types` set without `types_to_fetch` being None (and hence filtered_types being ignored).

Unless you want to put in the work to make the DictionaryCache more intelligent, I suspect you need to pull the `filtered_types` support back out of `_get_state_groups_from_groups`.

richvdh replied:
also: the `update the result` code below needs to take `filtered_types` into account.

ara4n replied:
so to check i understand: you're saying that `_get_state_for_groups` deliberately loads all state from the DB if we are querying state_key wildcards in order to fully populate the cache.  So none of the filtering gets applied to the DB, so the `filtered_types` stuff is irrelevant and will never apply.

I'm not convinced - surely if you call `_get_state_for_groups` with types of `[('m.room.member', '@matthew:matrix.org'), ('m.room.member', '@richvdh:sw1v.org')]`, there are no wildcards present, and so the request will go all the way through to the DB - and you'd need `filtered_types` to be set to `['m.room.member']` to get the rest of the state.

Or are you saying that because `filtered_types` acts as wildcard (i.e. "please return me all state outside this list"), we should go and blindly load all state out of the DB and then apply the filtering in the `update the result` phase?

ara4n replied:
i.e. something like https://gist.github.com/ara4n/55475cfe72c172e7df89b52eb20d25e3 ?

ara4n replied:
actually, i'll just push it.

richvdh replied:
> Or are you saying that because filtered_types acts as wildcard (i.e. "please return me all state outside this list"), we should go and blindly load all state out of the DB and then apply the filtering in the update the result phase?

yes, this.

ara4n commented on synapse/handlers/sync.py:
```diff
@@ -1472,11 +1541,22 @@ def _calculate_state(timeline_contains, timeline_start, previous, current):
+    tc_ids = set(e for e in timeline_contains.values())
+
+    # track the membership events in the state as of the start of the timeline
+    # so we can add them back in to the state if we're lazyloading.  We don't
```
have rewritten the comment again.

ara4n commented on synapse/handlers/sync.py:
```diff
@@ -1448,7 +1515,9 @@ def _action_has_highlight(actions):
 
-def _calculate_state(timeline_contains, timeline_start, previous, current):
+def _calculate_state(
+    timeline_contains, timeline_start, previous, current, lazy_load_members,
```
done.

ara4n commented on tests/storage/test_state.py:
```diff
@@ -0,0 +1,160 @@
+            (e1.type, e1.state_key): e1,
+            (e2.type, e2.state_key): e2,
+            (e3.type, e3.state_key): e3,
+        }, state)
```
done.

ara4n commented:
> I'm afraid that the more I look at how this has panned out the more I'm regretting encouraging the filtered_events thing. Somehow it's completely confusing.

Yes.

It feels like we've swapped an ugly but relatively intuitive wildcard pattern (that (None, None) means 'return everything else', just as (Type, None) means 'return everything that matches Type') for another subjectively ugly pattern (this meta-filter list which duplicates and can get out of sync with the keys of `types`).  I'm also unconvinced that performing contortions to avoid confusion over the edge case where `types` is `[]` is worthwhile when it makes the common case so confusing.

However, I think what we've ended up with here at least works, and is documented, so even if it's a bit confusing at first it's probably good enough.  I'd much much rather get this merged with a good-enough internal API than continue to try to get it perfect, especially given the pending stack of 6 PRs built on top of this one and the risk of conflicts the longer they stay in flight.

I've tried to incorporate your new wave of review; I believe the remaining pending points are:

https://github.com/matrix-org/synapse/pull/2970/files/efcdacad7d1b7f52f879179701c7e0d9b763511f#diff-6ee0c965f89a99e084e8500580e3ec17

and

https://github.com/matrix-org/synapse/pull/2970/files/efcdacad7d1b7f52f879179701c7e0d9b763511f#diff-6fee1701e980dbc9da57a5a108ee7b65

ara4n commented:
(and also a question over whether the sytests are robust enough, given some of the thinkos here haven't been picked up by them).

ara4n commented:
@richvdh ptal.

richvdh commented:
> It feels like we've swapped an ugly but relatively intuitive wildcard pattern (that (None, None) means 'return everything else', just as (Type, None) means 'return everything that matches Type') for another subjectively ugly pattern (this meta-filter list which duplicates and can get out of sync with the keys of types).

the problem was less about the ugliness (which might have been better solved with a "return_everything_else" boolean), and more about the base case of when types was `[]` (or `[(None, None)]`.

>   I'm also unconvinced that performing contortions to avoid confusion over the edge case where types is [] is worthwhile when it makes the common case so confusing.

Ok. I think that having the API behave completely differently when given an empty list was a massive footgun. At least we've got rid of that.

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -1545,9 +1548,12 @@ def _calculate_state(
-    # track the membership events in the state as of the start of the timeline
-    # so we can add them back in to the state if we're lazyloading.  We don't
-    # add them into state if they're already contained in the timeline.
+    # If we are lazyloading room members, we explicitly add the membership events
```
yay! I understand!

richvdh commented on synapse/handlers/sync.py:
```diff
@@ -1553,16 +1553,17 @@ def _calculate_state(
     # events by filtering them out of timeline_start, which has already been filtered
     # to only include membership events for the senders in the timeline.
+    # In practice, we can do this by removing them from the p_ids list.
+    # see https://github.com/matrix-org/synapse/pull/2970
```
I'm not sure a link that you have to reconstruct which proves that this implementation is equivalent to some other implementation that never landed is terribly helpful.

can we just say something like "In practice, we do this by removing them from p_ids, which is the list of things we know we have already sent to the client." ?

ara4n commented:
@richvdh ptal.

ara4n commented:
@richvdh ptal for a final time, i hope

richvdh commented on synapse/storage/state.py:
```diff
@@ -470,20 +500,33 @@ def _get_some_state_from_cache(self, group, types):
+
+            if (
+                state_key is None or
+                filtered_types is not None and typ not in filtered_types
```
lucky for you, `and` has higher precedence than `or`. I had to go and look it up though. Parens please.

ara4n replied:
done

richvdh commented on synapse/storage/state.py:
```diff
@@ -497,14 +540,20 @@ def _get_some_state_from_cache(self, group, types):
                 return True
             return False
 
-        got_all = is_all or not missing_types
```
can't we just write `is_all or (not missing_types and filtered_types is not None)` rather than special-casing over a particular bug elsewhere in the algorithm?

ara4n replied:
i find `is_all or (not missing_types and filtered_types is not None)` incredibly cryptic to reason about, to the extent i'm failing to convince myself it's even right.

For instance, if this is called with `types=[]` and `filtered_types=[whatever]`, then is_all could well be false (if the cache is empty), and missing_types would be falsey, and filtered_types would not be None... so got_all would be true, which is the wrong answer.

Which is why I very deliberately spelt out the special case we're handling here where `types=[]`, so we can't trust `missing_types`, which feels a lot easier to reason about.

richvdh replied:
ok, but the fact that you are special-casing the empty list makes *me* suspect that there are bugs elsewhere.

and sorry, it should have been `is_all or (not missing_types and filtered_types is None)`. Maybe it should be:

```
got_all = is_all
if not got_all:
    # the cache is incomplete. We may still have got all the results we need, if 
    # we don't have any wildcards in the match list.
    if not missing_types and filtered_types is None:
        got_all = True
```

richvdh commented on synapse/storage/state.py:
```diff
@@ -497,14 +540,20 @@ def _get_some_state_from_cache(self, group, types):
+        else:
+            got_all = is_all or not missing_types
 
         return {
```
worth noting that `missing_types` isn't actually used in the result. Suggest removing it; I think it might open some clearer options in how to implement this function

ara4n replied:
it feels like a useful thing for the function to be returning tbh, even if it isn't used, but given the filtered_types stuff means that we can't easily enumerate the types which are missing from the result, i've removed it.

richvdh commented on synapse/storage/state.py:
```diff
@@ -497,19 +539,24 @@ def _get_some_state_from_cache(self, group, types):
+        if not got_all:
+            # the cache is incomplete. We may still have got all the results we need, if
+            # we don't have any wildcards in the match list.
+            if not missing_types and filtered_types is None:
```
and now that we aren't returning `missing_types`, I wonder why we are bothering to build a set rather than just using a boolean. But let's just land the damn thing.

ara4n commented:
As per #synapse-dev:

```
* Matthew sighs
ironically GH didn't give me a squash merge button then
so i clicked merge anyway assuming it would prompt in a second phase
but it didn't
* Matthew wonders how to unpick that
given i assume a revert will make even more of a mess of the history
and force-pushing to develop will be even more antisocial.
```
